libs/sr/db/src/Database/DataManagers/TeamDataManager.ts

```diff
@@ -1,33 +1,59 @@
 import {
-  deepClone,
   objKeyMap,
   objMap,
+  pruneOrPadArray,
   range,
+  shallowCompareObj,
 } from '@genshin-optimizer/common/util'
-import type { RelicSlotKey } from '@genshin-optimizer/sr/consts'
-import { allRelicSlotKeys } from '@genshin-optimizer/sr/consts'
+import type { CharacterKey, RelicSlotKey } from '@genshin-optimizer/sr/consts'
+import {
+  allCharacterKeys,
+  allRelicSlotKeys,
+} from '@genshin-optimizer/sr/consts'
+import type { Member, Sheet, Tag } from '@genshin-optimizer/sr/formula'
 import type { ICachedRelic } from '../../Interfaces'
 import { DataManager } from '../DataManager'
 import type { SroDatabase } from '../Database'
+import { validateTag } from '../tagUtil'
 
 const buildTypeKeys = ['equipped', 'real', 'tc'] as const
 type BuildTypeKey = (typeof buildTypeKeys)[number]
-export type LoadoutMetadatum = {
-  loadoutId: string
+export type TeamMetadatum = {
+  characterKey: CharacterKey
+
   buildType: BuildTypeKey
   buildId: string
   buildTcId: string
   compare: boolean
   compareType: BuildTypeKey
   compareBuildId: string
   compareBuildTcId: string
+
+  optConfigId?: string
 }
 export interface Team {
   name: string
   description: string
 
-  loadoutMetadata: Array<LoadoutMetadatum | undefined>
   lastEdit: number
+
+  // frames, store data as a "sparse 2d array"
+  frames: Array<Tag>
+  conditionals: Array<{
+    sheet: Sheet
+    src: Member | 'all'
+    dst: Member | 'all'
+    condKey: string
+    condValues: number[] // should be the same length as `frames`
```

Comment:
Shouldnt this be stored inside a Frame obj then?

```typescript
frames: Array<Frame>
...
type Frame = {
  conditionals: Array<{
    sheet: Sheet
    src: Member | 'all'
    dst: Member | 'all'
    condKey: string
    condValue: number
}
bonusStats: Array<{
tag: Tag
value: number
}>
//etc

Reply:
Should save all the effort of keeping all the array lengths synchronized, which looks very prone to code breakage later

Reply:
I did consider that, the issue is that explodes the amount of data stored:
Frame obj: (4strings + 1 numbers)*conditionals*frames, vs 
sparce array: (4strings * conditionals) + 1 number*conditionals*frames

keeping array length the same is easier than syncing up and validating conditionals across an 2d array of conditionals.

Reply:
frame: `(4 strings + 1 number) * conditionals-in-frame * frames`

expanded frame: `1 number * conditionals-in-frame * frames + 4 strings * conditionals-in-frame * frames`

sparce: `1 number * conditionals-across-all-frames * frames + 4 strings * conditionals-across-all-frames`

so sparce gets worse if there are more conditionals across all frames, and there are a lot of frames. but should be pretty rare. Still a bit iffy on keeping the arrays all syncrhonized (e.g. adding frames in the middle, swaping frame positions, etc), but i'll wait until there are actual issues.